overleaf/track-changes
# [misc] migrate to the native mongo driver

### Description

For https://github.com/overleaf/issues/issues/2907
This PR is migrating the mongo driver from mongojs to mongodb -- the native driver.

The changes of overleaf/dev-environment#390 are included in the app commit.

#### Related Issues / PRs

https://github.com/overleaf/issues/issues/2907
Previous migration: https://github.com/overleaf/chat/pull/53 https://github.com/overleaf/chat/pull/54 https://github.com/overleaf/docstore/pull/66 (https://github.com/overleaf/contacts/pull/49 https://github.com/overleaf/document-updater/pull/144 https://github.com/overleaf/github-sync/pull/57 https://github.com/overleaf/notifications/pull/27 https://github.com/overleaf/project-history/pull/358 https://github.com/overleaf/spelling/pull/69 https://github.com/overleaf/third-party-datastore/pull/66 https://github.com/overleaf/third-party-references/pull/36 pending in review)

### Review

<details>
<summary> changed startup </summary>

The startup behavior has been changed in the chat service already https://github.com/overleaf/chat/pull/54 and is running with success in prod by now.

> This PR changes the startup of the service to not accept any http requests until after the mongo connection has been established. It also adds the correct way to handle connection setup errors, to catch and log them -- vs just letting them bubble up as unhandled PromiseRejections.
> 
> The web service can start without an established/working mongo connection, which is bad. Eventually we can port this snippet over.

</details>

I split the migration into multiple parts. There were quite a few `findAndModify` queries for which the call-sites ignore the `find` result. These have been changes to `updateOne` queries. The non-trivial changes to find queries have been split off from the main commit as well.

The callback signature for mongojs `findAndModify` and native driver `findOneAndUpdate` differ:
- `findAndModify`: `callback(error, doc)`
- `findOneAndUpdate`: `callback(error, {value: doc, ...})`

There are two locations in `PackManager.js` that needed changes in that regard. (Doing the `findAndModify -> updateOne` migration before hand helped in eliminating intermediate monkey-patching.) 

The `bson` package version has been bumped to the latest version in a separate commit.

#### Potential Impact

High. All mongo interactions: read/write project meta data and history contents.

#### Manual Testing Performed

We'll need a project with the old history backend.

- create a new project and note the projectId `5f3289a0e3bcb900263bf49b`  and main docId `5f3289a0e3bcb900263bf49c`

- delete the history link in mongo:

```
dev-env$ echo 'db.projects.updateOne({_id: ObjectId("5f3289a0e3bcb900263bf49b")}, {$unset: {overleaf: ""}})' | bin/mongo
```
- delete the history link in redis:
```
dev-env$ bin/exec redis redis-cli del "ProjectHistoryType:{5f3289a0e3bcb900263bf49c}"
```
- make changes in the editor
- open the project as a collaborator 'Alice' and let Alice perform some edits
- open the history tab
- see the history entries for the edits of  Alice and you in different colors
- revert to a previous revision
- see a banner for external changes
- see the previous revision in the editor
